src/middlewares/alert.ts

```diff
@@ -0,0 +1,75 @@
+import { create as createLogger } from '../common/log'
+const log = createLogger('alert-middleware')
+import { Middleware, MiddlewareCallback, Pipelines } from '../types/middleware'
+import * as IlpPacket from 'ilp-packet'
+
+const { T04_INSUFFICIENT_LIQUIDITY } = IlpPacket.Errors.codes
+
+export interface Alert {
+  id: number
+  accountId: string
+  message: string
+  count: number
+  createdAt: Date
+  updatedAt: Date
+}
+
+export default class AlertMiddleware implements Middleware {
+  private alerts: {[id: number]: Alert} = {}
+  private nextAlertId: number = Date.now()
+
+  async applyToPipelines (pipelines: Pipelines, accountId: string) {
+    pipelines.outgoingData.insertLast({
+      name: 'alert',
+      method: async (data: Buffer, next: MiddlewareCallback<Buffer, Buffer>) => {
+        const result = await next(data)
+        if (result[0] !== IlpPacket.Type.TYPE_ILP_REJECT) return result
+
+        const rejectPacket = IlpPacket.deserializeIlpReject(result)
+        if (rejectPacket.code !== T04_INSUFFICIENT_LIQUIDITY) return result
+
+        // The peer rejected a packet which, according to the local balance, should
+        // have succeeded. This can happen when our local connector owes the peer
+        // money but restarted before it was settled.
+        if (rejectPacket.message !== 'exceeded maximum balance.') return result
+
+        log.warn('generating alert for account=%s message="%s"', accountId, rejectPacket.message)
+        this.addAlert(accountId, rejectPacket.message)
```

Comment:
I think we should associate by the error's triggeredBy instead of by the accountId. There could be many routes that you reach through one peer, but different destinations on that route throw T04 errors

Reply:
I think I'll associate by the triple (accountId, triggeredBy, message)